pkg/controller/common/helpers.go

```diff
@@ -271,9 +296,107 @@ func IgnParseWrapper(rawIgn []byte) (ign2types.Config, error) {
+	}
+}
+
+// Function to remove duplicated files/units from a V2 MC, since the translator
+// (and ignition spec V3) does not allow for duplicated entries in one MC.
+// This should really not change the actual final behaviour, since it keeps
+// ordering into consideration and has contents from the highest alphanumeric
+// MC's final version of a file.
+// Note:
+// Append is not considered since we do not allow for appending
+// Units have one exception: dropins are concat'ed
+
+func removeIgnDuplicateFilesAndUnits(ignConfig ign2types.Config) ign2types.Config {
+
+	files := ignConfig.Storage.Files
+	units := ignConfig.Systemd.Units
+
+	filePathMap := map[string]bool{}
+	var outFiles []ign2types.File
+	for i := len(files) - 1; i >= 0; i-- {
+		// We do not actually support to other filesystems so we make the assumption that there is only 1 here
+		path := files[i].Path
+		if _, isDup := filePathMap[path]; isDup {
+			continue
+		}
+		outFiles = append(outFiles, files[i])
+		filePathMap[path] = true
+	}
+
+	unitNameMap := map[string]bool{}
+	var outUnits []ign2types.Unit
+	for i := len(units) - 1; i >= 0; i-- {
+		unitName := units[i].Name
+		if _, isDup := unitNameMap[unitName]; isDup {
+			// this is a duplicated unit by name, so let's check for the dropins and append them
+			if len(units[i].Dropins) > 0 {
+				for j := range outUnits {
+					if outUnits[j].Name == unitName {
+						// outUnits[j] is the highest priority entry with this unit name
+						// now loop over the new unit's dropins and append it if the name
+						// isn't duplicated in the existing unit's dropins
+						for _, newDropin := range units[i].Dropins {
+							hasExistingDropin := false
+							for _, existingDropins := range outUnits[j].Dropins {
+								if existingDropins.Name != newDropin.Name {
+									hasExistingDropin = true
+									break
+								}
+							}
+							if !hasExistingDropin {
+								outUnits[j].Dropins = append(outUnits[j].Dropins, newDropin)
+							}
+						}
+						continue
+					}
+				}
+				glog.Infof("Found duplicate unit %v, appending dropin section", unitName)
+			}
+			continue
+		}
+		outUnits = append(outUnits, units[i])
+		unitNameMap[unitName] = true
+	}
+
+	for _, unit := range outUnits {
+		glog.Infof("Remaining unit: %v", unit.Name)
+	}
+	// outFiles and outUnits should now have all duplication removed
+	ignConfig.Storage.Files = outFiles
+	ignConfig.Systemd.Units = outUnits
+
+	for _, unit := range ignConfig.Systemd.Units {
+		glog.Infof("Remaining unit in final conf: %v", unit.Name)
+	}
+
+	return ignConfig
+}
+
 // TranspileCoreOSConfigToIgn transpiles Fedora CoreOS config to ignition
-// internally it transpiles to Ign spec v3 config and translates to spec v2
-func TranspileCoreOSConfigToIgn(files, units []string) (*ign2types.Config, error) {
+// internally it transpiles to Ign spec v3 config
+func TranspileCoreOSConfigToIgn(files, units []string) (*ign3types.Config, error) {
```

Comment:
Now that we don't convert the FCCT templates to spec v2 anymore, we must ensure that the spec v3 config resulting from this function does not contain duplicate unit and file entries, i.e. they cannot be blindly appended like it has been done until now.

Reply:
wondering why we're not hitting this with `kubelet.service` here in CI already (there two templates for it: one that defines a common dropin, and one each for masters/workers that defines the contents)

Reply:
Edited so we use the updated fcct parser, and parse each file to igntion first and then merge them